File: examples/kotlin/src/test/kotlin/adr/ReplayTest.kt
package adr

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ReplayTest {

    @Test
    fun `re-folds a recorded timeline deterministically to a golden state and golden effects`() {
        val initial = State.initial(
            listOf(Ticket("4118", "refund not received", Priority.Normal, TicketStatus.Open)),
        )

        // a recorded session: the agent set priority and requested escalation;
        // the host then confirmed (a separate authority), which pages on-call.
        val recorded = listOf(
            RecordedStep(listOf(ToolResult.PrioritySet("4118", Priority.High)), 1000L),
            RecordedStep(listOf(ToolResult.EscalationRequested("4118", Actor.Agent)), 1001L),
            RecordedStep(listOf(ToolResult.EscalationConfirmed("4118", Actor.Human)), 1002L),
        )

        // `replay` folds twice and asserts the two are identical — it throws if not.
        val (state, effects) = initial.replay(recorded)

        val t = state.tickets.getValue("4118")
        assertEquals(Priority.High, t.priority)
        assertTrue(t.status is TicketStatus.Escalated)

        assertEquals(
            listOf(
                Effect.Log("priority", "4118", Priority.High, 1000L),
                Effect.PageOncall("4118", 1002L),
            ),
            effects,
        )
    }
}

File: examples/typescript/test/replay.test.ts
import { describe, it, expect } from "vitest";
import { initialState } from "../src/domain";
import { replay } from "../src/replay";
import type { RecordedStep } from "../src/fold";

describe("replay — the central claim", () => {
  it("re-folds a recorded timeline deterministically to a golden state and golden effects", () => {
    const initial = initialState([
      { id: "4118", body: "refund not received", priority: "Normal", status: { kind: "Open" } },
    ]);

    // a recorded session: the agent set priority and requested escalation;
    // the host then confirmed (a separate authority), which pages on-call.
    const recorded: RecordedStep[] = [
      { now: 1000, results: [{ kind: "PrioritySet", ticket: "4118", level: "High" }] },
      { now: 1001, results: [{ kind: "EscalationRequested", ticket: "4118", by: "Agent" }] },
      { now: 1002, results: [{ kind: "EscalationConfirmed", ticket: "4118", by: "Human" }] },
    ];

    // `replay` folds twice and asserts the two are identical — it throws if not.
    const { state, effects } = replay(initial, recorded);

    const t = state.tickets.get("4118")!;
    expect(t.priority).toBe("High");
    expect(t.status.kind).toBe("Escalated");

    expect(effects).toEqual([
      { kind: "Log", what: "priority", ticket: "4118", level: "High", at: 1000 },
      { kind: "PageOncall", ticket: "4118", at: 1002 },
    ]);
  });
});
